util/qc_executor.py

```diff
@@ -101,33 +102,36 @@ def qc_check(self, parameter, dataset):
             processid = os.fork()
             if processid == 0:
                 # child process
-                with os.fdopen(w, 'w') as w:
-                    try:
-                        os.close(r)
-                        # run the qc function
-                        results = getattr(module, function_name)(**local_qc_args.get(function_name))
-                        # convert the np.ndarray into a string for sending over pipe
-                        bytes = io.BytesIO()
-                        np.savez(bytes, x=results)
-                        results_string = bytes.getvalue()
-                        w.write(results_string)
-                    except (TypeError, ValueError) as e:
-                        log.exception('<%s> Failed to execute QC %s %r', self.request_id, function_name, e)
-                        w.write(EXCEPTION_MESSAGE)
-                    finally:
-                        # if w.flush() is not called (or w.close()), the parent will not get data for results_string 
-                        w.flush()
-                        # child process is done, don't let it stick around
-                        os._exit(0)
+                # don't use 'with os.fdopen' since we call os._exit(0) below
+                w = os.fdopen(w, 'w')
```

Comment:
I don't know if it's a good idea to override the original w with the return from fdopen. if fdopen fails or returns None or something does the original w get lost?

Reply:
I'll change the variables to prevent overwrite.

Reply:
After talking with John and giving it some thought, I agree that 'r' and 'w' should not be re purposed. I suspect you should also close the remaining pipes when you are done with them as they represent a potential resource leak. Each process closes the pipe it doesn't use; suggest also closing the pipe it does use after the end of each "with" block.